src/test/java/io/kamax/matrix/client/regular/MatrixHttpClientTest.java

```diff
@@ -39,6 +50,95 @@ public MatrixHttpClientTest() throws URISyntaxException {
     }
 
     @Test
+    public void loginWithDeviceId() throws URISyntaxException {
+        stubFor(post(urlEqualTo(loginUrl))
+                .withRequestBody(equalToJson("{\"type\": \"m.login.password\"," + //
+                        "\"user\": \"" + user.getLocalPart() + "\"," + //
+                        "\"password\": \"" + password + "\"," + //
+                        "\"device_id\": \"" + deviceId + "\"}"))
+                .willReturn(aResponse().withStatus(200)
+                        .withBody("{\"user_id\": \"" + user.getId() + "\"," + //
+                                "\"access_token\": \"" + testToken + "\"," + //
+                                "\"home_server\": \"" + new MatrixHomeserver(domain, baseUrl) + "\"," + //
+                                "\"device_id\": \"" + deviceId + "\"}")));
+
+        MatrixHomeserver hs = new MatrixHomeserver(domain, baseUrl);
+        MatrixClientContext context = new MatrixClientContext(hs, deviceId);
+        MatrixHttpClient client = new MatrixHttpClient(context);
+
+        MatrixPasswordLoginCredentials credentials = new MatrixPasswordLoginCredentials(user.getLocalPart(), password);
+        client.login(credentials);
+
+        assertEquals(deviceId, client.getDeviceId().get());
```

Comment:
As per spec:
> ID of the client device. If this does not correspond to a known client device, a new device will be created. The server will auto-generate a device_id if this is not specified.

While the test is valid in this specific case, it doesn't hold to the spec I believe. There is no guarantee you will have the same Device ID, which is a spec inconsistency I think, but best not assume this for now. Let's add a `TODO` instead to clarify with the Matrix devs.